### ISSUE TITLE
truffle deploy succeds when deploy dependencies, but truffle test doen't

### ISSUE THREAD

**Issue by Amxx** [trufflesuite/truffle#1795]
Context
======

I have contracts in `./contracts/` as well has dependencies in `./node_modules/someRepo/contracts`. The dependencies are already compiles and the artifacts are in `./node_modules/someRepo/build/contracts`

When I do `truffle deploy`, I see my contracts being compiled, and the deployment works for both the local contracts (`artifact.require("xxx")`) and the dependencies (`artifact.require("someRepo/yyy")`). Once the deployment has been performed I can run my tests with `truffle test`, `yyy.json` is also copied to `./build/contracts`

Issue
====

If I start fresh (`rm -r build`) and perform `truffle test`, truffle starts by compiling the local contract and then fails during the deployment with an error `Error: Could not find artifacts for someRepo/yyy from any sources`

Concerned version
===============

- for `5.0.0`, `5.0.1`, `5.0.2`, `5.0.3`, `5.0.4`, `5.0.5`, `5.0.7` I can deploy then test, but not test directly
- for `5.0.6` I cannot deploy nor test
- haven't tested with `4.x.x` (need solidity 0.5.x)

**Comment by CruzMolina** [comment]
Hey @Amxx! I've tried reproducing your error locally on a sample truffle project using `openzeppelin-solidity` dependency contracts, but so far things are working.

Could you share a truffle project repo w/ a setup similar to the one not working?

Tested on: latest Truffle v5 (`5.0.7`) & latest truffle v4 release (`4.1.15`).

**Comment by gnidan** [comment]
Doing some issue maintenance: @Amxx is this still an issue on the latest version? Please let us know if that PR above fixed it for you. Thanks!

**Comment by AleG94** [comment]
I'm currently having the same issue with Truffle 5.1.5. In the migration file I require artifacts from a dependency like:

`const ProxyFactory = artifacts.require('@openzeppelin/upgrades/ProxyFactory')`

If I run `truffle migrate` and then `truffle test` everything goes well but if I delete the build folder and I directly run `truffle test` (without migrate first) it will fail with error:

`Error: Could not find artifacts for @openzeppelin\upgrades\ProxyFactory from any sources`

**Comment by eggplantzzz** [comment]
@AleG94 Sorry for the delay in getting back to you! Can you confirm that this is still an issue on the latest Truffle (version 5.1.6)? And if it is, do you have a test repo that I could use to reproduce the problem?

**Comment by AleG94** [comment]
@eggplantzzz Yeah I can confirm it's still an issue in truffle 5.1.16.

I made an example repo for you, you can find it [here](https://github.com/AleG94/truffle-1795).

Running `truffle test` when the build folder does not exist will cause the error. Running `truffle migrate` and then `truffle test` works flawlessly.

**Comment by eggplantzzz** [comment]
Thanks @AleG94! I'll try and reproduce this today!

**Comment by AleG94** [comment]
@eggplantzzz is there any ETA? 
The issue is even blocking `solidity-coverage` from running.

**Comment by eggplantzzz** [comment]
Sorry @AleG94 I've been distracted as we have been working on preparing something for release soon. I will try and duplicate this today (this time for real)! :)

**Comment by eggplantzzz** [comment]
Ok so I understand why it is failing on your repo. You don't import that openzepellin contract into any of your other contracts and so when Truffle goes to compile, it doesn't know about it. When it goes to test, it can't resolve it because it assumes it has been compiled already. If you import it into one of your Solidity contracts, I think it would find it and compile it into the build folder.

I'm not to certain what we want to do about this. We'll have to have a discussion. Is it not a possibility to import it or somehow compile it beforehand?

**Comment by AleG94** [comment]
Yeah, it would be possible to import it into another contract but I'd have to create a `Workaround` contract just for that, so not a very clean solution. 

Since the contracts are already compiled in the npm module, I think it should be possible to deploy an imported contract 'as-is' directly from migrations. It would be a much cleaner way if truffle handled it this way.

For reference, in order to make `solidity-coverage` work, i had to bind a script to the `onCompileComplete` hook and manually copy `.json` files from `node_modules` to the folder when `solidity-coverage` saves the built artifacts.

**Comment by eggplantzzz** [comment]
Another hacky workaround is to require the JSON directly and then manually use `@truffle/contract` to convert it to a Truffle artifact. Something like
```
const ProxyFactory = require("<path>/<toJson>");
const contract = require("@truffle/contract");
const proxyFactory = contract(ProxyFactory);
```
Then you will be able to deploy it. I know, it's probably not the best but I thought I would leave it here anyways.

**Comment by gnidan** [comment]
It seems like the issue might be related to how `truffle test` hooks up @truffle/resolver. I think next steps for debugging this one are to look at when the [NPM resolver source](https://github.com/trufflesuite/truffle/blob/develop/packages/resolver/lib/sources/npm.ts) is invoked and see where it's going wrong. 

Not sure why this wouldn't work - looking at [prepareConfigAndRunTests](https://github.com/trufflesuite/truffle/blob/develop/packages/core/lib/commands/test/helpers.js#L48-L60), which I would expect to be the culprit, doesn't seem to affect this (the NPM resolver source uses the config `workingDirectory`, and that `prepareConfigAndRunTests` method only overrides the config `contractsBuildDirectory`.)

I've been in this code the past couple days; I will see if I can figure this out if I get some time tonight.

**Comment by eggplantzzz** [comment]
I believe this is resolved in version >5.1.35. Closing but if I'm mistaken we can re-open!
